### ISSUE TITLE
Are data weights really used?

### ISSUE THREAD

**Issue by allchemist** [deepchem/deepchem#2534]
Hello,

I have some datasets with missed values, and i set corresponding weights to zeros, but got no good training result. I looked inside the code and was unable to find out how weights are translated to loss functions.

Also i modified your example tox21_tensorgraph_graph_conv.py on tox21 dataset (which has missed values). The train_dataset.w matrix has 11k zeros, i changed them to ones, compared training result and got no difference. Seems data weights are not really used. Please have a look at passing data weights.

Thanks

**Comment by peastman** [comment]
Yes, they're really used!  For example, here is the line in KerasModel that scales the losses by the weights:

https://github.com/deepchem/deepchem/blob/f57af4842ec1d8644778d5602e03e5a435a8f4e9/deepchem/models/keras_model.py#L1227

This applies to standard loss functions.  If you supply your own custom loss to a model, it's up to you to handle weights.  GraphConvModel uses a standard loss, so that's not an issue:

https://github.com/deepchem/deepchem/blob/f57af4842ec1d8644778d5602e03e5a435a8f4e9/deepchem/models/graph_models.py#L956-L958

**Comment by allchemist** [comment]
Thanks for answer,

i found out that manual setting values to train_dataset.w does not really change them (after assigning train_dataset.w[0,0] = 0 it actually stays 1), so my experiment was not clear.

**Comment by allchemist** [comment]
By the way, can you share some ideas why multitask learning with missed values does not work? I tried different models and different datasets, training one data column works fine, with two or more columns - no progress in training.

**Comment by peastman** [comment]
Multitask learning should work fine with missing values.  Look at the ChEMBL examples.  It's a very sparse dataset with lots of tasks where most samples don't have data for most tasks.

**Comment by allchemist** [comment]
I found the problem - in ValidationCallback `model.evaluate` method is used, which did not use weights from validation dataset.
The `use_sample_weights` flag in `model.evaluate` returns error, so i wrote a workaround for myself (for binary weights case).

```
    # scores = model.evaluate(self.dataset, self.metrics)
    
    y_pred = model.predict(self.dataset).copy()

    zero_ids = np.vstack(np.where(self.dataset.w==0)).T
    for i,j in zero_ids:
          y_pred[i,j] = self.dataset.y[i,j]

    score = self.metrics[0].compute_metric(self.dataset.y, y_pred)
```

**Comment by peastman** [comment]
Thanks, good catch!  We should fix that to pass `use_sample_weights=True`.

**Comment by mgarort** [comment]
Hi @peastman could you please indicate the line where weights are used in PyTorch models? I cannot find it and I would like to understand how they are used.

My situation is similar to @allchemist 's : a dataset with missing values which should be weighted out.

Thank you.

**Comment by mgarort** [comment]
I found it myself!

https://github.com/deepchem/deepchem/blob/ee97ea5934b4ef853aaee4edf8e2ae04ff9ebef4/deepchem/models/torch_models/torch_model.py#L1194
